lib/ProductOpener/APITest.pm

```diff
@@ -166,17 +320,98 @@ For the example cited above this returns: "http://world-fr.openfoodfacts.localho
 
 =cut
 
-sub construct_test_url ($target, $prefix) {
-	my $link = "openfoodfacts.localhost";
-	# my $api_end = "/api/v2/search?";
-	my $api_end = "api/v2";
-
-	if (index($target, $api_end) != -1) {
+sub construct_test_url ($target, $prefix = "world") {
+	my $link = $TEST_MAIN_DOMAIN;
+	# no cgi inside url ? add display.pl
+	if (index($target, "cgi") == -1) {
 		$link .= "/cgi/display.pl?";
 	}
-
 	my $url = "http://${prefix}.${link}${target}";
 	return $url;
 }
 
+=head2 tail_log_start($log_path)
+
+Start monitoring a log file
+
+=head3 Arguments
+
+=head4 String $log_path
+
+Defaults to /var/log/apache2/log4perl.log
+
+=head3 Returns
+
+An object to pass to tail_log_read to read
+
+=cut
+
+sub tail_log_start ($log_path = "/var/log/apache2/log4perl.log") {
+	# we use nowait mode to avoid loosing time in test
+	# but beware, this means we will have to manually call checkpending()
+	# before reading
+	my $tail = File::Tail->new(name => $log_path, nowait => 1);
+	return $tail;
+}
+
+=head2 tail_log_read($tail)
+
+Return all content written to a log file since last check
+
+=head3 Arguments
+
+=head4 $tail
+
+Object returned by tail_log_start
+
+=head3 Returns
+
+Content as a string
+
+=cut
+
+sub tail_log_read ($tail) {
+	# we want to do a nowait read,
+	# but we bypass all the predict stuff from File::Tail
+	# by directly using checkpending
+	$tail->checkpending();
+	my @contents = ();
+	while (my $line = $tail->read()) {
+		push @contents, $line;
+	}
+	return join "", @contents;
+}
+
+=head2 mails_from_log($text)
+Retrieve mails in a log extract
+=cut
+
+sub mails_from_log ($text) {
```

Comment:
Being able to log emails and get them out of the logs so that we can test them is a very cool feature!